### PR TITLE
Add signatory labels to create and fetch in tree view

### LINKED ISSUE #9175: Feature request: In IDE script tx view add parties who initiate create
In the script runner tx view, when exercising a choice I see `'Alice' exercises...` but on `create` I don't have the equivalent party/parties who created the transaction. I would like to see `'Alice' creates...` or `'Alice', 'Bob' creates...`

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Resolves #9175
Changes create and fetch nodes in the tree viewer to show signatories
Formats as:
```
unknown creates ...
alice creates ...
alice and bob create ...
alice, bob and charlie create ...
```

Before:
```
TX 1 1970-01-01T00:00:00Z (Main:42:11)
  #1:0
  │   referenced by #2:1
  │   disclosed to (since): 'alice' (1), 'bob' (1) 
  └─> create Main:T
      with
        p = 'alice'; a = 2
```

After:
```
  TX 1 1970-01-01T00:00:00Z (Main:42:11)
  #1:0
  │   referenced by #2:1
  │   disclosed to (since): 'alice' (1), 'bob' (1) 
  └─> 'alice' and 'bob' create Main:T
              with
                p = 'alice'; a = 2
```

Exactly the same for fetch and exercise

CHANGELOG_BEGIN

- [Daml Studio] Update transaction tree view to show parties in more situations, with nicer formatting
See `#9175 <https://github.com/digital-asset/daml/issues/9175>`__.

CHANGELOG_END

### CODE REVIEW

**Review by akrmn — APPROVED** [review]
Thanks @samuel-williams-da!

I was a bit surprised this didn't affect any of the `EXPECTED.ledger` golden files, but then I remembered this is another piece of code that's duplicated between Haskell and Scala, and the `EXPECTED.ledger` files test the scala side. For reference, see #13735, which updates both `compiler/damlc/tests/src/DA/Test/ScriptService.hs` and `daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala`. 

I'm happy to merge this and let you work on the scala side in a separate PR, but if doing so please remember to keep #9175 open.

Also, since we mention the tree view in our documentation, there are a few other places to update for consistency; I think searching for `└─> create` + looking at the pictures under `docs/source/daml/intro/images/2_Scenario/*.png` should cover everything that needs to be updated, again see #13735 for reference; also fine if this is a separate PR.

**Review by remyhaemmerle-da — APPROVED** [review]
Nice. 
However I have to ask even I do not care much:
"No tests ?"

**Comment by samuel-williams-da** [comment]
Right right, I noticed the scala version of this while looking for existing tests, I can certainly update both, I think i just noticed a few little discrepancies between them and concluded they're probably not coupled anymore

I'll make those changes :)
As for tests on the _haskell_ side, given I didn't break any, i must conclude that the nodetree pretty printing is just not at all tested in haskell land, I could try set something up there, but maybe as a separate task

**Comment by akrmn** [comment]
> As for tests on the haskell side, given I didn't break any, i must conclude that the nodetree pretty printing is just not at all tested in haskell land

you're right here, sadly

> I could try set something up there, 

thanks! at least there's a suitable place for them, `compiler/scenario-service/client/test/DA/Daml/LF/PrettyScenarioSpec.hs`. On first look the tricky bit would seem getting all the setup to call `prettyNodeNode`, since it takes a `NodeNode`, which is a type generated from protobuf, and it runs in an [`M`](https://github.com/digital-asset/daml/blob/b8ffdb3443106ddd3c53511c1e53db0068690807/compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs#L54) context, so it needs a `MS.Map NodeId Node` and an `LF.World`

> but maybe as a separate task

👍

**Comment by samuel-williams-da** [comment]
Just redoing the screenshots, seems I don't have the same font in vscode, any idea what font (or even OS) these were taken on? or does it not matter? I ask as the difference is noticable
image

image

File: docs/source/daml/intro/7_Composing.rst [hunk 1]
```diff
@@ -150,21 +150,21 @@ The resulting transaction, with its two nested levels of consequences, can be se
...
           #14:6
           │   disclosed to (since): 'Alice' (14), 'Bob' (14), 'USD_Bank' (14)
-          └─> create Intro.Asset:Asset
-              with
-                issuer = 'USD_Bank';
-                owner = 'Bob';
-                symbol = "USD";
-                quantity = 100.0000000000;
-                observers = []
+          └─> 'Bob', 'USD_Bank' creates Intro.Asset:Asset
```
Comment (carrielaben-da) [inline docs/source/daml/intro/7_Composing.rst:161]:
```suggestion
          └─> 'Bob', 'USD_Bank' create Intro.Asset:Asset
```

**Review by carrielaben-da — APPROVED** [review]
Minor subject-verb agreement issue.

**Comment by carrielaben-da** [comment]
> Curious, would this not unalign the structure with exercise, in which the label is `X, Y exercises Z...`, rather than exercise? I was considering some simple logic to use `exercise/exercises` based on how many names before it: `Alice exercises X on Y` `Alice, Bob and Charcle exercise X on Y` Same for fetch + create, what do you think @carrielaben-da?

If this would not be too difficult, it would be better from a reader's perspective. I think it's a good idea.

**Comment by samuel-williams-da** [comment]
Should be fairly easy, and an opportunity to learn a little more scala ;)

**Comment by akrmn** [comment]
> Just redoing the screenshots, seems I don't have the same font in vscode, any idea what font (or even OS) these were taken on? or does it not matter? I ask as the difference is noticable

@samuel-williams-da, I took the previous ones on macOS using [Fira Code](https://github.com/tonsky/FiraCode); I think it's fine if you take them on a different OS, using the same font should get us close enough

**Comment by akrmn** [comment]
Still thinking about the screenshots, I'm wondering if the docs theme has an overarching style guide for things like this, would you know @carrielaben-da?

**Comment by samuel-williams-da** [comment]
I noticed `prettyEventInfo` in scala labels `fetch` as `ensure active`,  when neither `prettyPartialTransactionNode` or the haskell side of things does, is there a reason for this? Ideally they'd all be the same

File: docs/source/daml/daml-studio.rst [hunk 2]
```diff
@@ -319,9 +319,9 @@ choice 'Consume' of which he is not a controller
       TX #0 1970-01-01T00:00:00Z (unknown source)
       #0:0
       │   disclosed to (since): 'Alice' (#0), 'Bob' (#0)
-      └─> create ExerciseAuthFailure:Example
-          with
-            owner = 'Alice'; friend = 'Bob'
+      └─> 'Alice' creates ExerciseAuthFailure:Example
```
Comment (akrmn) [inline docs/source/daml/daml-studio.rst:322]: looking at this I noticed a difference between the haskell and scala pretty printers: the former quotes party values while the latter (sometimes) doesn't. I think it's easier to read with the quotes. In any case, this would be another PR.

Comment (samuel-williams-da) [reply docs/source/daml/daml-studio.rst:322]: I'm actually rewriting that bit rn, so its a trivial change I can add if we want it?

Comment (samuel-williams-da) [reply docs/source/daml/daml-studio.rst:322]: or do we think it needs more discussion

Comment (samuel-williams-da) [reply docs/source/daml/daml-studio.rst:322]: oh actually, theres a lot more cases than i thought, agree separate PR lol

File: compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs [hunk 3]
```diff
@@ -797,6 +806,18 @@ prettyNodeNode nn = do
         ppChildren <- prettyChildren node_RollbackChildren
         pure $ keyword_ "rollback" $$ ppChildren
 
+-- | Take a list of parties and the singular and multiple present tense verbs
+-- Depending on the count of parties, returns a prettified list of these elements, using @,@ and @and@, as well as the correct verb
+-- e.g. @[a, b, c] -> a, b and c@
+partiesAction :: V.Vector Party -> String -> String -> (Doc SyntaxClass, Doc SyntaxClass)
+partiesAction pv singular multiple = go $ mapV prettyParty pv
```
Comment (akrmn) [inline compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs:813]: more a style thing than anything else; `go` makes me think there's going to be recursion, I think here a simple `case` on `mapV prettyParty pv` would be easier to follow.

Comment (samuel-williams-da) [reply compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs:813]: yeah thats fair, i was considering calling it aux, but i realise it doesn't really need to be its own thing

File: compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs [hunk 4]
```diff
@@ -797,6 +806,18 @@ prettyNodeNode nn = do
...
+  where
+    go [] = (text "No-one/unknown", keyword_ singular)
+    go [p] = (p, keyword_ singular)
+    go ps = (fcommasep init <-> keyword_ "and" <-> head last, keyword_ multiple)
+      where
+        (init, last) = splitAt (length ps - 1) ps
```
Comment (akrmn) [inline compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs:819]: here I'd use [`unsnoc`](https://hackage.haskell.org/package/extra-1.7.12/docs/Data-List-Extra.html#v:unsnoc) since this [already depends on `extra`](https://github.com/digital-asset/daml/blob/akrmn/dynamic-exercise/compiler/scenario-service/client/BUILD.bazel#L21).

Comment (samuel-williams-da) [reply compiler/scenario-service/client/src/DA/Daml/LF/PrettyScenario.hs:819]: I can see that, though that Maybe could get messy without a fromJust

File: docs/source/daml/daml-studio.rst (hunk 2 above)
Comment (akrmn) [reply docs/source/daml/daml-studio.rst:322]: > or do we think it needs more discussion

not really, it's more about keeping the changes easier to follow

**Comment by akrmn** [comment]
> I noticed `prettyEventInfo` in scala labels `fetch` as `ensure active`,  when neither `prettyPartialTransactionNode` or the haskell side of things does, is there a reason for this? Ideally they'd all be the same

interesting, I hadn't seen this one before. I do agree they should all match. It seems the `ensure active` _and_ `fetch` lines both come from the 'open-sourcing daml' commit so I don't know what caused the divergence, but it makes sense to me to keep `fetch`.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala [hunk 5]
```diff
@@ -197,9 +197,13 @@ private[lf] object Pretty {
       case Node.Rollback(_) =>
         text("rollback")
       case create: Node.Create =>
-        "create" &: prettyContractInst(create.coinst)
+        intercalate(text(", "), create.signatories.map(p => text(p))) &
+          text("creates") &
```
Comment (akrmn) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:201]: I think the only thing missing in this PR is getting these to agree with the number of parties, right?

Comment (samuel-williams-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:201]: Yep, I'm working on aligning the scala part to haskell changes rn

**Comment by carrielaben-da** [comment]
> Still thinking about the screenshots, I'm wondering if the docs team has an overarching style guide for things like this, would you know @carrielaben-da?

As far as I know we do not, but I will reach out to the UX team and see if they have guidance.

**Comment by carrielaben-da** [comment]
@akrmn Per discussion with Jennifer we shouldn't let consistency of the screenshots be a blocker here, but the UX team will work to get a standard in place for consistency in the future.

**Comment by akrmn** [comment]
@carrielaben-da thank you very much for looking into this :)

**Comment by samuel-williams-da** [comment]
Looked into it, running fire code in my browser seems to give me a different font completely, the `0`s have a line through them
I'll just merge as is, and once we have a standard, we can fix it all up
